modules/frontend/search_streaming.go

```diff
@@ -95,17 +96,74 @@ func (p *diffSearchProgress) finalResult() *shardedSearchResults {
 	return p.progress.result()
 }
 
+type multiProgress struct {
+	mu       sync.Mutex
+	progress []*atomic.Pointer[*diffSearchProgress]
+}
+
+func newMultiProgress() *multiProgress {
+	return &multiProgress{
+		progress: make([]*atomic.Pointer[*diffSearchProgress], 0),
+	}
+}
+
+func (mp *multiProgress) Add(p *atomic.Pointer[*diffSearchProgress]) {
+	mp.mu.Lock()
+	defer mp.mu.Unlock()
+
+	mp.progress = append(mp.progress, p)
+}
+
+// Snapshot returns a safe copy of progress list. Use this function to get a copy of progress list
+// and use it to do further processing on the progress objects.
+func (mp *multiProgress) Snapshot() []*atomic.Pointer[*diffSearchProgress] {
+	mp.mu.Lock()
+	defer mp.mu.Unlock()
+
+	copyList := make([]*atomic.Pointer[*diffSearchProgress], len(mp.progress))
+	copy(copyList, mp.progress)
+	return copyList
+}
+
+func (mp *multiProgress) finalResults(ctx context.Context) *shardedSearchResults {
+	mp.mu.Lock()
+	defer mp.mu.Unlock()
+
+	if len(mp.progress) == 1 {
+		// single tenant query, only one progress object.
+		p := *mp.progress[0].Load()
+		return p.finalResult()
+	}
+
+	// multi-tenant query, merge result from all progress objects.
+	// note: newSearchProgress is used to combine results from multiple progress objects, don't use it for anything else.
+	comb := newSearchProgress(ctx, 0, 0, 0, 0)
```

Comment:
should we be using `combiner.NewSearch` here? i'm honestly not sure. wdyt?

Reply:
ah, i see now, the multi tenant middleware is used to split the requests but they're recombined in the streaming multi progress.